packages/openapi3/src/openapi.ts

```diff
@@ -450,95 +458,21 @@ function createOAPIEmitter(program: Program, options: OpenAPIEmitterOptions) {
-              codes.push(String(prop.type.value));
-            }
-          } else if (prop.type.kind === "Union") {
-            for (const option of prop.type.options) {
-              if (option.kind === "String") {
-                if (validStatusCode(option.value, option)) {
-                  codes.push(option.value);
-                }
-              } else if (option.kind === "Number") {
-                if (validStatusCode(String(option.value), option)) {
-                  codes.push(String(option.value));
-                }
-              } else {
-                reportDiagnostic(program, { code: "status-code-invalid", target: prop });
-              }
-            }
-          } else {
-            reportDiagnostic(program, { code: "status-code-invalid", target: prop });
-          }
+          codes.push(...getStatusCodes(program, prop));
         }
       }
     }
-    return codes;
-  }
-
-  // Check status code value: 3 digits, 1 digit + "XX", or default
-  // Issue a diagnostic if not valid
-  function validStatusCode(code: string, entity: Type): boolean {
-    // regex for three character status codes:
-    // - starts with 1-5
-    // - last two digits are numeric or "X"
-    const statusCodePatten = /[1-5][0-9X][0-9X]/;
-    if (code.match(statusCodePatten) || code === "default") {
-      return true;
+    if (isDefaultResponse(program, responseModel)) {
+      if (codes.length > 0) {
+        reportDiagnostic(program, {
+          code: "status-code-in-default-response",
+          target: responseModel,
+        });
+      } else {
+        codes.push("default");
+      }
     }
-    reportDiagnostic(program, {
-      code: "status-code-invalid",
-      target: entity,
-      messageId: "value",
-    });
-    return false;
-  }
-
-  // Note: these descriptions come from https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
-  function getDescriptionForStatusCode(statusCode: string) {
-    switch (statusCode) {
-      case "2XX":
-        return "Successful";
-      case "200":
-        return "Ok";
-      case "201":
-        "Created";
-      case "202":
-        return "Accepted";
-      case "204":
-        return "No Content";
-      case "3XX":
-        return "Redirection";
-      case "301":
-        return "Moved Permanently";
-      case "304":
-        return "Not Modified";
-      case "4XX":
-        return "Client Error";
-      case "400":
-        return "Bad Request";
-      case "401":
-        return "Unauthorized";
-      case "403":
-        return "Forbidden";
-      case "404":
-        return "Not Found";
-      case "409":
-        return "Conflict";
-      case "412":
-        return "Precondition Failed";
-      case "5XX":
-        return "Server Error";
-      case "default":
-        return "An unexpected error response";
-    }
-    // We might want to throw here -- rather than giving some default
-    return "A successful response";
```

Comment:
I moved the status code description method to the http library and improved the handling of codes not explicitly listed.